### ISSUE TITLE
Not possible to allow daemonset pods with hostnetwork=true to reach another pod in the same namespace

### ISSUE THREAD

**Issue by Cajga** [aws/eks-anywhere#4765]
**What happened**:
We want to use "default-deny" network policy so we installed the cluster with policyEnforcementMode: "always" config option in Cilium. This will enable policy enforcement by default on both Ingress and Egress traffic so without a network policy which allows a specific traffic, everything is blocked.

Our CSI solution ([Astra Trident](https://docs.netapp.com/us-en/trident/) but it is irrelevant for now) is working in the following way:
* it has a controller pod which is deployed through a deployment
* it has pods on all nodes which are handling the volume management created through a daemonset (with hostnetwork=true)
* when a daemonset pod start up, it "registers the node" to the controller
* all of these pods are running in the same namespace (called trident)

As we have default-deny, we have to write a network policy to allow the traffic from the daemonset pods to the controller pod.
We tried first with a very strict policy which did not work. So we took another approach this time and started from a widely open policy and tried to narrow it down step by step.

**1) Opening up both ingress and egress works fine**
Apply the following policy
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-all-ingress-egress
  namespace: trident
spec:
  podSelector: {}
  egress:
  - {}
  ingress:
  - {}
  policyTypes:
  - Egress
  - Ingress
```

The daemonset pods can talk to the controller and came up fine:
```
$ kubectl get pods -n trident |grep csi
trident-csi-6566868bcb-rr8sb       6/6     Running   6 (7d22h ago)   7d22h
trident-csi-8lfwc                  2/2     Running   0               46s
trident-csi-ncwjt                  2/2     Running   0               23s
trident-csi-tbvcq                  2/2     Running   0               46s
trident-csi-vctbt                  2/2     Running   0               46s
$ kubectl logs -n trident trident-csi-8lfwc trident-main |grep registrat
time="2023-01-26T12:45:06Z" level=info msg="Updated Trident controller with node registration." node=NODENAME1 requestID=e6786bc8-0de2-4900-a0a3-310e25cad87f requestSource=Internal
```

**2) As a first step in the restriction path, we tried to only allow pods in the same namespace**
Apply the following policy (after removed the allow all policy from above)
```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: allow-csi-daemonset-communication
  namespace: trident
spec:
  podSelector: {}
  egress:
    - {}
  ingress:
    # Allow all intra-namespace traffic
    - from:
      - podSelector: {}
  policyTypes:
    - Ingress
    - Egress
```

With this policy, we end up in a very interesting status. Only the daemonset pod which is running on the same node as the controller is able to reach the controller (that is the only CSI pod which could startup both containers):
```
$ kubectl get pods -n trident  -o wide|grep csi
trident-csi-4cn4j                  1/2     Running   0               5m15s   IPADDRNODE04    node004   <none>           <none>
trident-csi-6566868bcb-rr8sb       6/6     Running   6 (7d22h ago)   7d22h   IPADDRNODE14     node014   <none>           <none>
trident-csi-8qhkm                  1/2     Running   0               5m15s   IPADDRNODE11   node011   <none>           <none>
trident-csi-brxlw                  1/2     Running   0               5m14s   IPADDRNODE12   node012   <none>           <none>
trident-csi-qhw4n                  2/2     Running   0               5m21s   IPADDRNODE14    node014   <none>           <none>

$ kubectl logs -n trident trident-csi-8qhkm trident-main |grep registrat|head -n 1
time="2023-01-26T12:48:38Z" level=warning msg="Could not update Trident controller with node registration, will retry." error="could not log into the Trident CSI Controller: error communicating with Trident CSI Controller; Put \"https://100.81.227.206:34571/trident/v1/node/node011\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)" increment=10.920863391s requestID=9b04f1d9-cfe5-4742-a583-4923659b5cc3 requestSource=Internal

``` 

Possibly, this is due to the fact that in cilium, the ingress is always allowed from the host itself. We came to this conclusion through checking the allowed identities of the controller pod. We run these commands inside the cilium pod where the controller pod running to get the allowed identities and verify that identity 1 is the host itself:
```
# cilium endpoint list|grep app=controller.csi.trident.netapp.io
156        Enabled            Enabled           56111      k8s:app=controller.csi.trident.netapp.io                                                      100.80.3.65    ready   
# cilium endpoint get 156|grep -A 4 allowed-ingress-identities 
          "allowed-ingress-identities": [
            44854,
            56111,
            1
          ],
# cilium identity get 1
ID   LABELS
1    reserved:host
```

**So, allowing all the traffic inside the namespace is still not enough for the daemonset pods (except the one on the same host) to reach the controller. Seems there is no way to allow such traffic without opening up everything**
Note: Based on our research, it seems, pods with hostnetwork=true are not covered by network policies as their network is not provided by the CNI itslef. Some CNI implementation (like Cilium) do cover these "special" pods but only through their extended, CRD based network policies.

**What you expected to happen**:
Be able to follow least-privilege, zero-trust principles and lock down the EKS-A cluster while allow traffic from daemonset pods (with hostNetwork=true) to other pods in the same namespace.

**How to reproduce it (as minimally and precisely as possible)**:
Install a cluster with policyEnforcementMode: "always" and try to install trident through [trident operator helm chart](https://artifacthub.io/packages/helm/netapp-trident/trident-operator) (no need to have netapp appliance to test this!)

**Anything else we need to know?**:

**Environment**:
- EKS Anywhere Release: v0.13.1
- EKS Distro Release: 1.24

**Comment by ddjjia** [comment]
related issue: https://github.com/aws/eks-anywhere/issues/4658

**Comment by chrisdoherty4** [comment]
> I believe with that with https://github.com/aws/eks-anywhere/issues/4658 fixed, it should be possible to construct a KNP to meet the user requirement. This would include:
>
> * "default-deny" policy applied to all endpoints
> * a KNP for the namespace: trident matching all endpoints and allowing traffic from/to CIDRs covering node IPs
>
> The downside of this is that this would allow communication with any process on the nodes but at least it should restrict access for pods in other namespaces

**Comment by chrisdoherty4** [comment]
Resolved by #5030
